docs: cover interim bases (#971)
canonical/rockcraft: docs/explanation/bases.rst

Reviewer: https://canonical-ubuntu-documentation-library--971.com.readthedocs.build/rockcraft/971/explanation/bases/

:relatedlinks: https://ubuntu.com/about/release-cycle, [Ubuntu&#32;image&#32;on&#32;Docker&#32;Hub](https://hub.docker.com/_/ubuntu), [Ubuntu&#32;image&#32;on&#32;Amazon&#32;ECR](https://gallery.ecr.aws/ubuntu/ubuntu), [Scratch&#32;image&#32;on&#32;Docker&#32;Hub](https://hub.docker.com/_/scratch)

> 

Reviewer 2: how come we need to add `&#32;` for spaces?
Also it kind of feels like this extension should support multi-line entries? Seems weird that you should define multiple links on a single line

> 

Reviewer 3: I've already flagged both the ASCII whitespace and multi-line support as improvements to be made. They're high on my list of things to be fixed once I update `canonical-sphinx` to use the reworked extensions. Setting up the tests has unfortunately left me little time to implement major improvements.

> 

Author: I've documented the issues. I won't make this a mainstay of our docs until after these are resolved. This instance here is just a proving ground.

https://github.com/canonical/canonical-sphinx-extensions/issues/75
https://github.com/canonical/canonical-sphinx-extensions/issues/76
https://github.com/canonical/canonical-sphinx-extensions/issues/77
https://github.com/canonical/canonical-sphinx-extensions/issues/78

.. _bases_explanation:

Bases
=====

Every rock is built from a *base*, which defines the baseline system that the rock's

> contents are layered on. The systems are always major Ubuntu releases. An empty, or

Author: The [Ubuntu packaging guide](https://canonical-ubuntu-packaging-guide.readthedocs-hosted.com/en/latest/explanation/archive/#series) notes that this term has caused confusion.

Actually, looking into it further, the images have stable and edge channels, and are brought up to some kind of update. @tigarmo Do they bring in point releases?

Regardless, I think this wording is just as true and less ambiguous: d05096d

also available for lean images.

The base is declared in the project file by the :literalref:`base <rockcraft_yaml_base>`
key.

Ubuntu bases
------------

When building an Ubuntu base, Rockcraft pulls the corresponding Ubuntu image from a

> container registry. These Ubuntu images are the same as the ones hosted in registries

Author: We have future plans for a different registry, so the wording is general for the time being.

like Docker Hub or Amazon Elastic Container Registry.

The built rock contains any software and utilities that are present in the Ubuntu image.

Setting an Ubuntu base for a rock is especially useful when the goal is to build a rock
that can be used for general-purpose environments, such as a development working space.

LTS and interim bases

> ---------------------

Author: Agreed. I wanted a flatter hierarchy, but it works better with the subsection. d1b34d8

Ubuntu bases are divided into *LTS* and *interim* bases. An LTS base contains an Ubuntu
LTS release and has a 10-year support window. An interim base contains an interim Ubuntu
release and is supported for nine months.

Interim bases are less stable than LTS bases because they contain experimental features

> and have shorter lifespans. They are most suitable for container-based testing the

Reviewer: Not sure we really want to say interim bases are less stable (also... not my experience ;-p). Maybe just point out that interim bases only have a 9 month lifespan and are thus best suited to preparing for the next LTS release?

Ubuntu release itself, using it as a proving ground for upcoming features and
compatibility.

Bare bases
----------

As the name suggests, a *bare* base results in a rock with no baseline system. This
definition is similar to the scratch Docker image, with the exception that a rock *can

> never be completely empty*, as it must always include a minimum baseline with
> :ref:`pebble_explanation_page` and :ref:`additional metadata <what-sets-rocks-apart>`.

Reviewer: ```suggestion
never be completely empty*, as it must always include :ref:`pebble_explanation_page` 
and some :ref:`additional metadata <what-sets-rocks-apart>`.
```

> The bare base is extremely useful when the goal is to build a lean chiselled rock.

Author: d1b34d8

Such rocks are typically preferred for production environments due to their optimized

> deployment efficiency and reduced attack surface. The combination of the bare base with

Reviewer: I know this isn't rewording the previous text, but I'm not sure I agree with this. I've historically switched from very small Alpine bases for Docker to much bigger Ubuntu-based container images specifically because the shared layers resulted in a saving of storage space and memory.

Bare bases are very useful though when the rock is the product (e.g. a rock containing a database server) rather than just a means of deployment (e.g various microservices for a single website).

Reviewer 2: Assuming the exception is with the "optimized deployment efficiency", then I see Alex's point but I don't think this is a change that needs to happen on this PR specifically (it's something we should bring up with Cris, he wrote the original text I believe).

:ref:`Chisel <chisel_explanation>` slices will result in a minimalist container image
that meets the production environment's criteria while retaining its functionality.

> When using the bare base, the rock will *still derive from Ubuntu*. As such, Rockcraft

Author: I was trying not to reuse "base" to overload the meaning.

If it's not strictly true, should we rephrase this to instead simply say that rocks need an Ubuntu base during assembly? I find "bare bases are still based on Ubuntu" to be needlessly complex to try and work around.
